evaluation_service/src/main.rs

```diff
@@ -1,80 +1,583 @@
+use std::io;
+use std::path::PathBuf;
+
 use protos::service::evaluation::{evaluation_server::*, *};
 use protos::utils::*;
 use tonic::{transport::*, Request, Response, Status};
+use utils::storage::FsStorageHelper;
+
+const ROOT_PATH: &str = "/evaluation_files";
+const SERIALIZED_EXTENSION: &str = "ser";
+const USER_SCORING_FILE_NAME: &str = "user_scoring";
+const PROBLEMS_FOLDER_NAME: &str = "problems";
+const TESTCASES_FOLDER_NAME: &str = "testcases";
+const EVALUATION_FILES_FOLDER_NAME: &str = "files";
+const INPUT_FILE_NAME: &str = "input";
+const OUTPUT_FILE_NAME: &str = "output";
+const IO_EXTENSION: &str = "txt";
+const PROBLEM_METADATA_FILE_NAME: &str = "metadata";
+
+fn internal_error<T>(e: T) -> Status
+where
+    T: core::fmt::Debug + core::fmt::Display,
+{
+    Status::internal(format!("{:?}", e))
+}
+
+fn not_found_error<T>(e: T) -> Status
+where
+    T: core::fmt::Debug + core::fmt::Display,
+{
+    Status::not_found(format!("{:?}", e))
+}
+
+fn not_found_io_error<T>(e: T) -> io::Error
+where
+    T: core::fmt::Debug + core::fmt::Display,
+{
+    io::Error::new(io::ErrorKind::NotFound, format!("{:?}", e))
+}
 
 #[derive(Debug)]
-pub struct EvaluationService {}
+pub struct EvaluationService {
+    storage: FsStorageHelper,
+}
+
+impl EvaluationService {
+    fn get_problem_folder_from_id(&self, problem_id: u64) -> io::Result<PathBuf> {
+        self.storage
+            .search_item(None, PROBLEMS_FOLDER_NAME, None)
+            .and_then(|op| op.ok_or_else(|| not_found_io_error("Problems folder not found")))
+            .and_then(|path| {
+                self.storage
+                    .search_item(Some(&path), &problem_id.to_string(), None)
+            })
+            .and_then(|op| {
+                op.ok_or_else(|| {
+                    not_found_io_error(&format!("Problem not found [id: {}]", problem_id))
+                })
+            })
+    }
+}
 
 #[tonic::async_trait]
 impl Evaluation for EvaluationService {
     async fn get_user_scoring(
         &self,
         _request: Request<GetUserScoringRequest>,
     ) -> Result<Response<GetUserScoringResponse>, Status> {
-        todo!()
+        self.storage
+            .search_item(None, USER_SCORING_FILE_NAME, Some(SERIALIZED_EXTENSION))
+            .map_err(not_found_error)
```

Comment:
Is this really a not_found_error? would that not return Ok(None)?

Reply:
Ah yes, I'll fix asap.